### PR TITLE
Rholang: Normalizer: Add support for match

### PR DESCRIPTION
**Pull request by KentShikama**
Note this allows duplicate case patterns. The following is
still valid.

```
for (x <- @Nil) {
    match x {
        case (a|b) => Nil
        case (a|b) => Nil
    }
}
```

### CODE REVIEW

**Comment by KentShikama** [comment]
This resolves https://rchain.atlassian.net/browse/RHOL-140 .

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 1]
```diff
@@ -337,6 +337,32 @@ object ProcNormalizeMatcher {
         ProcVisitOutputs(input.par.prepend(foldedNew), bodyResult.knownFree)
       }
 
+      case p: PMatch => {
+        import scala.collection.JavaConverters._
+
+        val value = normalizeMatch(p.proc_, input)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Please use something like targetResult.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 2]
```diff
@@ -337,6 +337,32 @@ object ProcNormalizeMatcher {
...
+          case ci: CaseImpl => (ci.proc_1, ci.proc_2)
+          case _            => throw new Error("Unexpected Case implementation.")
+        }
+
+        val initAcc = (List[Tuple2[Par, Par]](), input.knownFree)
+        val casesResult = (initAcc /: cases) { (acc, caseImpl) =>
+          caseImpl match {
+            case (pattern, caseBody) => {
+              val patternResult =
+                normalizeMatch(pattern, ProcVisitInputs(Par(), input.env, input.knownFree))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: The pattern should be normalized with an empty free map.

**Comment by guardbotmk3** [comment]
The test cases should include free variables. One test should compile and one should fail because the pattern has the same free variable twice.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 1 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 2 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

**Comment by KentShikama** [comment]
@guardbotmk3 Ok ready for another review.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 3]
```diff
@@ -348,6 +348,35 @@ object ProcNormalizeMatcher {
...
+          case _            => throw new Error("Unexpected Case implementation.")
+        }
+
+        val initAcc = (List[Tuple2[Par, Par]](), input.knownFree)
+        val casesResult = (initAcc /: cases) { (acc, caseImpl) =>
+          caseImpl match {
+            case (pattern, caseBody) => {
+              val patternResult =
+                normalizeMatch(pattern,
+                               ProcVisitInputs(Par(), input.env, DebruijnLevelMap[VarSort]()))
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Patterns also cannot reference the environment.

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Done

**Pull request merged**